file-events/src/test/groovy/net/rubygrapefruit/platform/file/HardLinkFileEventFunctionsTest.groovy

```diff
@@ -0,0 +1,188 @@
+            (MAC_OS): [change(MODIFIED, rootDir)],
+            (LINUX): [change(CREATED, link)]
+        )
+    }
+
+    def "does not detect hard link created outside watched hierarchy"() {
+        given:
+        def watched = new File(rootDir, "watched")
+        watched.mkdirs()
+        def target = new File(watched, "target")
+        target.createNewFile()
+        def link = new File(rootDir, "link")
+        startWatcher(watched)
+
+        when:
+        createHardLink(link, target)
+
+        then:
+        // macOS reports a change to the directory containing the original file
+        expectEvents byPlatform(
+            (MAC_OS): [change(MODIFIED, watched)],
+            (OTHERWISE): []
+        )
+    }
+
+    def "can detect hard link modified"() {
+        given:
+        def target = new File(rootDir, "target")
+        target.createNewFile()
+        def link = new File(rootDir, "link")
+        createHardLink(link, target)
+        startWatcher(rootDir)
+
+        when:
+        link.text = "modified"
+
+        then:
+        // On Windows we seem to be getting multiple MODIFIED events
+        // On Linux we _sometimes_ seem to be getting multiple events
+        expectEvents byPlatform(
+            (MAC_OS): [change(MODIFIED, link)],
+            (WINDOWS): [change(MODIFIED, link)] * 2,
+            (LINUX): [change(MODIFIED, link), optionalChange(MODIFIED, link)]
+        )
+    }
+
+    def "does not detect hard link modified outside watched hierarchy"() {
+        given:
+        def watched = new File(rootDir, "watched")
+        watched.mkdirs()
+        def target = new File(watched, "target")
+        target.createNewFile()
+        def link = new File(rootDir, "link")
+        createHardLink(link, target)
+        startWatcher(watched)
+
+        when:
+        link.text = "modified"
+
+        then:
+        expectNoEvents()
+
+        when:
+        target.text = "modified2"
+
+        then:
+        // On Windows we seem to be getting multiple MODIFIED events
+        // On Linux we _sometimes_ seem to be getting multiple events
+        expectEvents byPlatform(
+            (MAC_OS): [change(MODIFIED, target)],
+            (WINDOWS): [change(MODIFIED, target)] * 2,
+            (LINUX): [change(MODIFIED, target), optionalChange(MODIFIED, target)]
+        )
+    }
+
+    def "can detect hard link removed"() {
+        given:
+        def target = new File(rootDir, "target")
+        target.createNewFile()
+        def link = new File(rootDir, "link")
+        createHardLink(link, target)
+        startWatcher(rootDir)
+
+        when:
+        link.delete()
+
+        then:
+        // Windows sometimes reports a modification before the removal
+        expectEvents byPlatform(
+            (WINDOWS): [optionalChange(MODIFIED, link), change(REMOVED, link)],
+            (OTHERWISE): [change(REMOVED, link)]
+        )
+    }
+
+    def "does not detect hard link removed outside watched hierarchy"() {
```

Comment:
Same here: If I remove a link pointing to a file in the watched directory, why would that even matter? Is this just for coverage so we don't see any strange events?

Reply:
Exactly, just as above.